src/ipa_pc/mod.rs

```diff
@@ -64,22 +66,46 @@ impl<G: AffineCurve, D: Digest> InnerProductArgPC<G, D> {
         Ok(comm)
     }
 
-    #[inline]
-    fn inner_product(l: &[G::ScalarField], r: &[G::ScalarField]) -> G::ScalarField {
-        l.par_iter().zip(r).map(|(li, ri)| *li * ri).sum()
-    }
-
     /// Complete semantic checks on `ck`.
     #[inline]
     pub fn check_key(ck: &CommitterKey<G>, max_degree: usize) -> bool {
-        let pp = <Self as PolynomialCommitment<G::ScalarField>>::setup(max_degree).unwrap();
+        let pp = <Self as PolynomialCommitment<G>>::setup(max_degree).unwrap();
         ck.is_valid() && &pp.hash == &ck.hash
     }
 
-    /// Computes an opening proof of multiple check polynomials with a corresponding
-    /// commitment GFin opened at point.
-    /// No segmentation here: Bullet Polys are at most as big as the committer key.
-    pub fn open_check_polys<'a>(
+    /// Perform a dlog reduction step as described in BCMS20
+    fn round_reduce(
+        round_challenge: G::ScalarField,
+        round_challenge_inv: G::ScalarField,
+        c_l: &mut [G::ScalarField],
+        c_r: &[G::ScalarField],
+        z_l: &mut [G::ScalarField],
+        z_r: &[G::ScalarField],
+        k_l: &mut [G::Projective],
+        k_r: &[G],
+    ) {
+
+        c_l.par_iter_mut()
+            .zip(c_r)
+            .for_each(|(c_l, c_r)| *c_l += &(round_challenge_inv * c_r));
+
+        z_l.par_iter_mut()
+            .zip(z_r)
+            .for_each(|(z_l, z_r)| *z_l += &(round_challenge * z_r));
+
+        k_l.par_iter_mut()
+            .zip(k_r)
```

Comment:
We can improve the prover performance a lot, if we use projective representation of k_r instead of k_l. Then the mul is carried out entirely without field inversions.

Reply:
Sorry, my mistake: the current implementation of mul for affine points already uses internally projective representation together with mixed addition.

Reply:
And once again, rewind: mul for AffineCurve outputs projective representation. So no improvement by what I proposed.